### PR TITLE
New PyEventPlot widget

### PR DESCRIPTION
**Pull request by mcb64**
I was having a lot of difficulty with the PyDMScatterPlot widget.  In particular, it really wasn't doing a very good job of pairing matching X and Y values, so even when attempting to plot the same value on *both* X and Y, there were a lot of mismatched values.

This is a pretty shameless reworking of PyDMScatterPlot, except it accepts a single channel, which is assumed to be an event-built array.  Indices into the array are used to select the X and Y values for each curve.

This has the same bugs as PyDMScatterPlot, most notably that changing some parameters inside designer seems to break the display, as something in C++ seems to be getting deleted prematurely by clear().

### CODE REVIEW

**Review by jbellister-slac — COMMENTED** [review]
Thanks for the PR! Would be interested to get a better idea of what your use case for this is - and specifically curious what the PV you are plotting is/looks like just to get a better sense of the general purpose use of the plot.

I will also take a look at the designer/clear() error issue, if there are any other bugs you were running into please let us know, happy to fix on our end!

**Comment by mcb64** [comment]
> Thanks for the PR! Would be interested to get a better idea of what your use case for this is - and specifically curious what 
> the PV you are plotting is/looks like just to get a better sense of the general purpose use of the plot.

Sure.  This all started with the PCDS timetool... on each shot, the DAQ writes an array of 8 doubles to a PV called TTALL, and users wanted to add additional elements to this array and do various correlation plots with the elements in the array.  So I started looking into how to use the PyDMScatterPlot here and to be honest, it gave me fits.  While I could use a calc data plugin to pick out elements to plot... things just didn't seem to work well.  I didn't have constant access to the DAQ, and needed to append elements to this array anyway, so I wrote an aSub EPICS event builder.  As a test, I made an array of three doubles: GDET:FEE:241:ENRC twice, and GDET:FEE:242:ENRC once.  Then trying out PyDMScatterPlot, I tried to pull out and plot 241 vs. itself... and I couldn't get this to work using any redraw mode.  Giving up on the calc plugin and directly plotting GDET:FEE:241:ENRC vs. itself only approximately worked as well... yes, most of the points were on a straight line in every case, but a significant fraction weren't.

At this point, I decided that since my data was already in an event-built array, why rip it apart and hope PyDM can put it back together?  So I decided to modify the widget to plot this directly.  The simplest thing seemed to be to duplicate it and edit it appropriately: hence PyDMEventPlot.

It's late Friday night, but I will address your other issues on Tuesday. :-)

File: pydm/widgets/eventplot.py [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+        new_data: numpy.ndarray
+            A new array of values.
+        """
+        if new_data is None:
+            return
+        if len(new_data) <= self.x_idx or len(new_data) <= self.y_idx:
+            return
+        self.data_buffer = np.roll(self.data_buffer, -1)
+        self.data_buffer[0, -1] = new_data[self.x_idx]
+        self.data_buffer[1, -1] = new_data[self.y_idx]
```
[the remark this answers is not in the document]
Comment (mcb64) [reply pydm/widgets/eventplot.py:112]: I've added a test to return if either x_idx or y_idx is None.

File: pydm/widgets/eventplot.py [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+        init_y_channels, they both must have the same length.
+        If a single x channel was specified, and a list of y channels are
+        specified, all y channels will be plotted against the same x channel.
+    init_y_indices: optional
+        init_y_channels can be a string with the address for a channel,
+        or a list of strings, each containing an address for a channel.
+        If a list is specified for both init_x_channels and init_y_channels,
+        they both must have the same length.
+        If a single x channel was specified, and a list of y channels are
+        specified, all y channels will be plotted against the same x channel.
```
[the remark this answers is not in the document]
Comment (mcb64) [reply pydm/widgets/eventplot.py]: Yeah, I've reworded this completely.  I wasn't terribly careful correcting this comment.

File: pydm/widgets/eventplot.py [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+        if lineWidth is not None:
+            plot_opts['lineWidth'] = lineWidth
+        curve = EventPlotCurveItem(addr=channel,
+                                         y_idx=y_idx,
+                                         x_idx=x_idx,
+                                         name=name,
+                                         color=color,
+                                         yAxisName=yAxisName,
+                                         bufferSizeChannelAddress=bufferSizeChannelAddress,
+                                         **plot_opts)
```
[the remark this answers is not in the document]
Comment (mcb64) [reply pydm/widgets/eventplot.py]: Yep.

**Review by jbellister-slac — APPROVED** [review]
Appreciate the additional information. Code changes and comments look good. And I think others may find this useful as well as the waveform can be in any format. Thanks!

**Pull request merged**